File: yuanfang-html-image/scripts/extract.py
#!/usr/bin/env python3
"""Extract content from a URL or text file, output JSON for render.js"""

import sys
import json
import re
import requests
from lxml import html as lxml_html
from urllib.parse import urlparse

def extract_from_url(url):
    headers = {
        'User-Agent': 'Mozilla/5.0 (compatible; yuanfang-html-image/1.0)',
        'Accept': 'text/html,application/xhtml+xml',
    }
    try:
        resp = requests.get(url, headers=headers, timeout=15)
        resp.raise_for_status()
        resp.encoding = resp.apparent_encoding or 'utf-8'
    except Exception as e:
        return {'title': f'Error fetching URL', 'content': str(e), 'source': url, 'points': []}

    tree = lxml_html.fromstring(resp.content)

    title = ''
    og_title = tree.xpath("//meta[@property='og:title']/@content")
    if og_title:
        title = og_title[0]
    else:
        t = tree.xpath("//title/text()")
        if t:
            title = t[0].strip()
    title = clean_text(title)

    desc = ''
    og_desc = tree.xpath("//meta[@property='og:description']/@content")
    if og_desc:
        desc = og_desc[0]
    else:
        meta_desc = tree.xpath("//meta[@name='description']/@content")
        if meta_desc:
            desc = meta_desc[0]
    desc = clean_text(desc)

    body_texts = []
    for tag in ['h1', 'h2', 'h3', 'p']:
        elements = tree.xpath(f'//{tag}[not(ancestor::nav)][not(ancestor::header)][not(ancestor::footer)]/text()')
        for el in elements[:30]:
            t = clean_text(el)
            if len(t) > 15:
                body_texts.append(t)

    body = ' '.join(body_texts[:8]) if body_texts else desc

    return {
        'title': title or urlparse(url).netloc,
        'content': body or desc,
        'source': url,
        'points': [],
    }

def extract_from_text(text):
    lines = [l.strip() for l in text.strip().split('\n') if l.strip()]
    if not lines:
        return {'title': '', 'content': '', 'source': '', 'points': []}
    title = lines[0]
    content_lines = []
    points = []
    for line in lines[1:]:
        if line.startswith('-') or line.startswith('•') or line.startswith('*'):
            points.append(line.lstrip('-•* '))
        else:
            content_lines.append(line)
    return {
        'title': title,
        'content': ' '.join(content_lines) if content_lines else '',
        'source': '',
        'points': points,
    }

def clean_text(t):
    t = re.sub(r'\s+', ' ', t or '')
    return t.strip()[:200]

def main():
    if len(sys.argv) < 2:
        print(json.dumps({'title': '', 'content': '', 'source': '', 'points': []}))
        return

    input_arg = sys.argv[1]

    if input_arg.startswith('http://') or input_arg.startswith('https://'):
        result = extract_from_url(input_arg)
    elif input_arg == '--file' and len(sys.argv) > 2:
        with open(sys.argv[2], 'r') as f:
            text = f.read()
        result = extract_from_text(text)
    else:
        result = extract_from_text(input_arg)

    print(json.dumps(result, ensure_ascii=False))

if __name__ == '__main__':
    main()

File: yuanfang-html-image/scripts/extract.js
#!/usr/bin/env node
const fs = require('fs');

function htmlDecode(s) {
  if (!s) return s;
  return s
    .replace(/&amp;/g, '&')
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&quot;/g, '"')
    .replace(/&#39;/g, "'")
    .replace(/&nbsp;/g, ' ')
    .replace(/&#x([0-9a-fA-F]+);/g, (_, h) => String.fromCharCode(parseInt(h, 16)))
    .replace(/&#(\d+);/g, (_, d) => String.fromCharCode(parseInt(d, 10)));
}

function getMetaContent(html, name) {
  let m = html.match(new RegExp(
    `<meta\\s+name=["']${name}["']\\s+content=["']([^"']*)["']`, 'i'));
  if (m) return htmlDecode(m[1]);
  m = html.match(new RegExp(
    `<meta\\s+content=["']([^"']*)["']\\s+name=["']${name}["']`, 'i'));
  if (m) return htmlDecode(m[1]);
  m = html.match(new RegExp(
    `<meta\\s+property=["']og:${name}["']\\s+content=["']([^"']*)["']`, 'i'));
  if (m) return htmlDecode(m[1]);
  m = html.match(new RegExp(
    `<meta\\s+content=["']([^"']*)["']\\s+property=["']og:${name}["']`, 'i'));
  if (m) return htmlDecode(m[1]);
  return null;
}

function getLinkHref(html, rel) {
  const patterns = [
    new RegExp(`<link[^>]+rel=["']${rel}["'][^>]+href=["']([^"']+)["']`, 'i'),
    new RegExp(`<link[^>]+href=["']([^"']+)["'][^>]+rel=["']${rel}["']`, 'i'),
  ];
  for (const re of patterns) {
    const m = html.match(re);
    if (m) return m[1];
  }
  return null;
}

function stripTags(html) {
  return html.replace(/<[^>]+>/g, ' ').replace(/\s+/g, ' ').trim();
}

function extractTitle(html) {
  return getMetaContent(html, 'og:title')
    || getMetaContent(html, 'twitter:title')
    || getMetaContent(html, 'title')
    || (() => {
        const m = html.match(/<h1[^>]*>([\s\S]*?)<\/h1>/i);
        return m ? stripTags(m[1]) : null;
      })()
    || '';
}

function extractBody(html) {
  return getMetaContent(html, 'og:description')
    || getMetaContent(html, 'description')
    || getMetaContent(html, 'twitter:description')
    || (() => {
        const m = html.match(/<p[^>]*>([\s\S]*?)<\/p>/i);
        return m ? stripTags(m[1]) : '';
      })();
}

function extractBrand(html, url) {
  return getMetaContent(html, 'site_name')
    || getMetaContent(html, 'application-name')
    || (url ? new URL(url).hostname.replace(/^www\./, '') : '');
}

function extractBrandImage(html) {
  return getMetaContent(html, 'image')
    || getLinkHref(html, 'apple-touch-icon')
    || getLinkHref(html, 'apple-touch-icon-precomposed')
    || getLinkHref(html, 'icon');
}

function extractPoints(html) {
  const points = [];
  const listMatch = html.match(/<(ul|ol)[^>]*>([\s\S]*?)<\/\1>/i);
  if (listMatch) {
    const items = listMatch[2].matchAll(/<li[^>]*>([\s\S]*?)<\/li>/gi);
    for (const m of items) {
      const t = stripTags(m[1]);
      if (t && t.length < 80) points.push(t);
      if (points.length >= 3) break;
    }
  }
  if (points.length === 0) {
    const headings = html.matchAll(/<h[23][^>]*>([\s\S]*?)<\/h[23]>/gi);
    for (const m of headings) {
      const t = stripTags(m[1]);
      if (t && t.length < 60) points.push(t);
      if (points.length >= 3) break;
    }
  }
  return points;
}

function extractFromHtml(html, url) {
  return {
    title: stripTags(extractTitle(html) || ''),
    body: extractBody(html) || '',
    points: extractPoints(html),
    brand: extractBrand(html, url),
    brandImage: extractBrandImage(html) || undefined,
  };
}

function extractFromText(text) {
  const lines = text.split('\n').map(l => l.trim()).filter(Boolean);
  const title = lines[0] || '';
  const body = lines.find((l, i) => i > 0 && !l.match(/^[-•*]\s/) && !l.match(/^公众号[\s:：]/)) || '';
  const points = lines
    .filter(l => l.match(/^[-•*]\s/))
    .map(l => l.replace(/^[-•*]\s+/, ''))
    .slice(0, 3);
  let brand = '';
  const brandMatch = text.match(/(?:公众号|来源|来自|from|by|via)[\s:：]+([^\n]+)/i);
  if (brandMatch) brand = brandMatch[1].trim();
  return { title, body, points, brand };
}

async function extractFromUrl(url) {
  const res = await fetch(url, {
    headers: { 'User-Agent': 'yuanfang-skills/0.1 (+https://github.com/yuanfang)' },
    redirect: 'follow',
  });
  if (!res.ok) throw new Error(`HTTP ${res.status} fetching ${url}`);
  const html = await res.text();
  return extractFromHtml(html, url);
}

function parseArgs(argv) {
  const args = argv.slice(2);
  if (args.length === 0) {
    throw new Error('Usage: extract.js <url> | --text "..." | --file <path>');
  }
  if (args[0] === '--text') {
    return { mode: 'text', text: args.slice(1).join(' ') };
  }
  if (args[0] === '--file') {
    if (!args[1]) throw new Error('--file requires a path');
    return { mode: 'file', path: args[1] };
  }
  return { mode: 'url', url: args[0] };
}

async function main() {
  const args = parseArgs(process.argv);
  let result;

  if (args.mode === 'url') {
    result = await extractFromUrl(args.url);
  } else if (args.mode === 'text') {
    result = extractFromText(args.text);
  } else if (args.mode === 'file') {
    const text = fs.readFileSync(args.path, 'utf8');
    result = extractFromText(text);
  }

  process.stdout.write(JSON.stringify(result, null, 2) + '\n');
}

module.exports = {
  htmlDecode,
  getMetaContent,
  getLinkHref,
  stripTags,
  extractTitle,
  extractBody,
  extractBrand,
  extractBrandImage,
  extractPoints,
  extractFromHtml,
  extractFromText,
  extractFromUrl,
  parseArgs,
};

if (require.main === module) {
  main().catch(err => { console.error(err.message); process.exit(1); });
}
